cmd/deploy/translate.go

```diff
@@ -0,0 +1,209 @@
+)
+
+func translateBody(b []byte, name string) ([]byte, error) {
+	var body map[string]json.RawMessage
+	if err := json.Unmarshal(b, &body); err != nil {
+		return nil, fmt.Errorf("could not unmarshal request: %s", err)
+	}
+
+	if err := translateMetadata(body, name); err != nil {
+		return nil, err
+	}
+
+	var typeMeta metav1.TypeMeta
+	if err := json.Unmarshal(b, &typeMeta); err != nil {
+		return nil, fmt.Errorf("could not process resource's type: %s", err)
+	}
+
+	switch typeMeta.Kind {
+	case "Deployment":
+		if err := translateDeploymentSpec(body, name); err != nil {
+			return nil, err
+		}
+	case "StatefulSet":
+		if err := translateStatefulSetSpec(body, name); err != nil {
+			return nil, err
+		}
+	case "Job":
+		if err := translateJobSpec(body, name); err != nil {
+			return nil, err
+		}
+	case "CronJob":
+		if err := translateCronJobSpec(body, name); err != nil {
+			return nil, err
+		}
+	case "DaemonSet":
+		if err := translateDaemonSetSpec(body, name); err != nil {
+			return nil, err
+		}
+	case "ReplicationController":
+		if err := translateReplicationControllerSpec(body, name); err != nil {
+			return nil, err
+		}
+	case "ReplicaSet":
+		if err := translateReplicaSetSpec(body, name); err != nil {
+			return nil, err
+		}
+	}
+
+	return json.Marshal(body)
+}
+
+func translateMetadata(body map[string]json.RawMessage, name string) error {
+	m, ok := body["metadata"]
+	if !ok {
+		return fmt.Errorf("request body doesn't have metadata field")
+	}
+
+	var metadata metav1.ObjectMeta
+	if err := json.Unmarshal(m, &metadata); err != nil {
+		return fmt.Errorf("could not process resource's metadata: %s", err)
+	}
+
+	if metadata.Labels == nil {
+		metadata.Labels = map[string]string{}
+	}
+	metadata.Labels[model.DeployedByLabel] = name
+
+	if metadata.Annotations == nil {
+		metadata.Annotations = map[string]string{}
+	}
+	if utils.IsOktetoRepo() {
+		metadata.Annotations[model.OktetoSampleAnnotation] = "true"
+	}
+
+	metadataAsByte, err := json.Marshal(metadata)
+	if err != nil {
+		return fmt.Errorf("could not process resource's metadata: %s", err)
+	}
+
+	body["metadata"] = metadataAsByte
+
+	return nil
+}
+
+func translateDeploymentSpec(body map[string]json.RawMessage, name string) error {
+	var spec appsv1.DeploymentSpec
+	if err := json.Unmarshal(body["spec"], &spec); err != nil {
+		return fmt.Errorf("could not process deployment spec: %s", err)
+	}
+	spec.Template.Labels[model.DeployedByLabel] = name
+	specAsByte, err := json.Marshal(spec)
+	if err != nil {
+		return fmt.Errorf("could not process deployment's spec: %s", err)
+	}
+	body["spec"] = specAsByte
+	return nil
+}
+
+func translateStatefulSetSpec(body map[string]json.RawMessage, name string) error {
+	var spec appsv1.StatefulSetSpec
+	if err := json.Unmarshal(body["spec"], &spec); err != nil {
+		return fmt.Errorf("could not process statefulset spec: %s", err)
+	}
+	spec.Template.Labels[model.DeployedByLabel] = name
+	for i := range spec.VolumeClaimTemplates {
```

Comment:
@ifbyol with this, we will already know which volumes where created by statefulsets. I think we had logic to detect them on `okteto destroy -v`, could you point me to where that logic is? I would like to add a comment there to remove the logic at some point.

Reply:
@pchico83 this is the function we use for what you mention: https://github.com/okteto/okteto/blob/master/pkg/k8s/namespaces/namespace.go#L142

Reply:
@ifbyol I won't modify the volumeclaimtempalte yet, modifying an existing statefulset volumeclaimtemplate is unsupported

Reply:
@pchico83 other option could be to add that label in the volumes only on POST operations, not sure if there could be any operation in which could provoke an error

Reply:
@ifbyol good point, but I think we will always need to handle both use cases, because people can deploy things with kubectl directly

Reply:
@pchico83 but even using directly `kubectl`, the API operation under the hood to create the statefulset would be a POST, right? I mean, Kubernetes API is rest, so we could check which operations are POST in the proxy to include the label or not. We are currently handling only POST and PUT: https://github.com/okteto/okteto/blob/master/cmd/deploy/proxy.go#L210

Reply:
yes, but I could create the statefulset with volumeclaimtemplates with `kubectl` and then, we will always need the code to remove those volumes manually

Reply:
oh, yes, to delete the volume, we will need the code that does it manually, but starting to include these labels on creation will help us to group those volumes with the pipelines they belong when they are sleeping for example

Reply:
makes sense, I will update the logic :+1:

Reply:
still problematic, I could create the sfs with `okteto deploy` and then a `kubectl apply` would fail...

Reply:
of, so let's not include it then

Reply:
sounds good. auto modifying this field is risky...